src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,315 @@
+
+export type StreamNameParts<T extends StreamType = StreamType> = {
+  streamType: T;
+  entityId: string;
+};
+
+export type StreamToProject<EventType extends Event> = {
+  streamName: StreamName;
+  streamType: StreamType;
+  entityId: string;
+  streamVersion: number;
+  events: EventType[];
+};
+
+export interface EventStream {
+  streamName: string;
+  events: Array<{
+    _id: ObjectId;
+    type: string;
+    data: string;
+    metadata: string;
+  }>;
+  createdAt: Date;
+  updatedAt: Date;
+}
+export type EventStreamEvent = EventStream['events'][number];
+
+export interface MongoDBConnectionOptions {
+  connectionString: string;
+  database: string;
+  collection?: string;
+}
+
+class EventStoreClass implements EventStore<number> {
+  private readonly collection: Collection<EventStream>;
+
+  constructor(collection: typeof this.collection) {
+    this.collection = collection;
+  }
+
+  async readStream<EventType extends Event>(
+    streamName: StreamName,
+    options?: ReadStreamOptions<number>,
+  ): Promise<Exclude<ReadStreamResult<EventType, number>, null>> {
+    const expectedStreamVersion = options?.expectedStreamVersion;
+    const stream = await this.collection.findOne({
+      streamName: { $eq: streamName },
+    });
+
+    if (!stream) {
+      return {
+        events: [],
+        currentStreamVersion: MongoDBEventStoreDefaultStreamVersion,
+        streamExists: false,
+      };
+    }
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    const formattedEvents = stream.events
+      .slice(0, maxEventIndex(expectedStreamVersion))
+      .map(this.parseEvent<EventType>(streamName));
+
+    return {
+      events: formattedEvents,
+      currentStreamVersion: stream.events.length,
+      streamExists: true,
+    };
+  }
+
+  async aggregateStream<State, EventType extends Event>(
+    streamName: StreamName,
+    options: AggregateStreamOptions<State, EventType, number>,
+  ): Promise<AggregateStreamResult<State, number>> {
+    const stream = await this.readStream<EventType>(streamName, options?.read);
+    const state = stream.events.reduce(options.evolve, options.initialState());
+    return {
+      state,
+      currentStreamVersion: stream.currentStreamVersion,
+      streamExists: stream.streamExists,
+    };
+  }
+
+  async appendToStream<EventType extends Event>(
+    streamName: StreamName,
+    events: EventType[],
+    options?: AppendToStreamOptions<number> & {
+      /**
+       * These will be ran after a the events have been successfully appended to
+       * the stream. `appendToStream` will return after every projection is completed.
+       */
+      projections?: Array<
+        (stream: StreamToProject<EventType>) => void | Promise<void>
+      >;
+      /**
+       * Same as `options.projections` but this will run asynchronously.
+       */
+      asyncProjections?: Array<
```

Comment:
I think I've seen some other ways to handle registering projections in your talks so let me know if this isn't _the way_.

Reply:
I think that we'll need to find a good way for that. Either using Mongo change streams to trigger those projections or having some sort collection with the outboxed messages. Have you thought about it already?